First draft supporting configuration directory for the linux agent

mrud opened the pull request:
This will potentially break tests but a first draft of implementing a conf.d concept for the linux agent.

Please have a look @vilda / @StephenHynes7

mrud commented:
this also breaks the concept of reinit which should probably go away....

vilda commented:
reinit is being used, removal of this command would cause problems

vilda commented:
Otherwise looks ok to me. Will need to understand/describe what's the behavior if you specify same parameters in multiple configurations. Order does not seem to be guaranteed (fix as simple as adding sort).

mrud commented:
The ordering is based on the names of the files, e.g. `01` is before `02`. Files can overwrite each other.

mrud commented:
The problem with re-init is that it will write a new configuration, meaning it will combine the individual configurations into one

vilda commented on README.md:
```diff
@@ -110,6 +110,11 @@ The agent stores configuration in `~/.le/config` for ordinary users and in
+The agent supports a configuration directory which is, per default
+`~/.le/conf.d/` for ordinary users and `/etc/le/conf.d` for root (daemon).
+Files matching `*.config` will be loaded and used as configuration from the
+config directory. The order of the loaded files depends on the naming. 
```
"The order of the loaded files depends on the naming." is not specific enough. Better would be "Configuration files are loaded according to their alphabetical ordering."

vilda commented on README.md:
```diff
@@ -130,6 +135,18 @@ In the main section, `user-key` (account key) which identifies account, and
+
+The configuration directory is set per default to `~/.le/conf.d/` for ordinary
+users and to `/etc/le/conf.d` for root. To change the directory the `include_directory`
+can be specified in the main section to adjust the directory:
```
Better wording. I would say: To change the configuration directory, specify the include_directory in the main section as follows:

I would use "include" as the name instead if that's ok?

vilda commented:
LGTM with the exception of reinit. Should we document in more detail that reinit may break settings? Or disable it in case user is using conf.d (i.e. specifying include_directory parameter)?

mrud commented:
I personally would just go with it as it is. I think using reinit and conf.d is probably exclusive as one would be used for managing individual systems and one would be more likely to be used for configuration management

micahhausler commented:
Just to chime in, what is the reasoning on enforcing files end with `.config`? The format is technically `.ini` and in the [python world](https://docs.python.org/2/library/configparser.html#ConfigParser.RawConfigParser.read), the convention for file extensions for ConfigParser files is `.cfg`. The main config file `/etc/le/config` doesn't even have an extension, so why not just match on all files regardless of the extension in `/etc/le/conf.d`?

Also a quick note about the code, the `.read()` method on ConfigParsers accepts multiple files so the loop is unnecessary as you can just pass in the full list.

mrud commented:
.config was used as the main configuration file is named config. I wanted to be in sync with it.

I wanted to have debug out but I'll change it

mrud commented:
@vilda  + @micahhausler  i think i addressed now all concerns, running reinit will not include data from included directories, changed the matching to `*.cfg` and am no longer printing each file

vilda commented:
Thanks @mrud @micahhausler for pull requests provided and the discussion. I'll merge these two PR together and deploy a new release.

Looking around, the most commonly used configuration extensions in `/etc/*.d` is "no extension" or `.conf`. I'm more inclined towards `.conf` since it avoids common issues with ghost files created by editors (`*~`, `*.tmp`) and packaging systems (`*.dpkg-dist`).

micahhausler commented:
@vilda now that you point out the ghost files, an extension makes a lot of sense and either `.conf` or .`cfg` is fine by me. 

One more thought is that editors with filetype highlighting will correctly interpret the `.cfg` format, but thats really just a nicety rather than affecting functionality.

vilda commented:
Merged via #136
